### PR TITLE
More vec_bcd_ppc.h:

### PR DESCRIPTION
**Pull request by munroesj52**
Clean up of Extend/carry support for BCD add/sub.
Fix a bug with borrow.
Added BCD compare for predicate and bool
New functions; vec_bcdcfz vec_bcdctz
Add shift and round, shift unsigned, Truncate, and
Unsigned truncate.
Latentcy estimates for new functions.
Add Doxygen exmaple of extented multiply.

	* src/vec_bcd_ppc.h: Update Doxygen description.
	[\section bcd128_details_0_0]: Added Some details of BCD
	computation.
	[\subsection bcd128_extended_0_1]: Added Preferred sign, zone,
	and zero.
	[\subsection bcd128_extended_0_2]: Was a section.
	Corrected some of the code examples.
	[\subsubsection bcd128_extended_0_2_1; Added Vector BCD
	Multiply Quadword example.
	(vbBCD_): New typedef for vector bool for BCD.
	(vec_bcdsub): Forward ref prototype.
	(vec_bcdaddcsq): Code improvements, reduced latency.
	(vec_bcdaddecsq): Code improvements, reduced latency.
	(vec_bcdcfz): New function.
	(vec_bcdcmp_eqsq, vec_bcdcmp_gesq, vec_bcdcmp_gtsq,
	vec_bcdcmp_lesq, vec_bcdcmp_ltsq, vec_bcdcmp_nesq):
	New function.
	(vec_bcdcmpeq. vec_bcdcmpge, vec_bcdcmpgt, vec_bcdcmple,
	vec_bcdcmplt, vec_bcdcmpne): New function.
	(vec_bcdctz, vec_bcdsetsgn, vec_bcdslqi, vec_bcdsluqi,
	vec_bcdsr, vec_bcdsrrqi): New function.
	(vec_bcdsubcsq): Code improvements, reduced latency.
	(vec_bcdsubecsq): Code improvements, reduced latency.
	(vec_bcdsubesqm): Code improvements.
	(vec_bcdtrunc, vec_bcdtruncqi, vec_bcdutrunc, vec_bcdutruncqi):
	New function.
	(vec_cbcdaddcsq, vec_cbcdaddecsq):  Code improvements.
	(vec_cbcdsubcsq): New function.

	*src/vec_common_ppc.h (VEC_HW_L_DWH): define for truncate
	digit count hword.

Signed-off-by: Steven Munroe <munroesj52@gmail.com>

### CODE REVIEW

File: src/vec_bcd_ppc.h [hunk 1]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
   r_h = vec_bcdaddesqm (a_h, b_h, c_l)
  * \endcode
  *
+ * \todo The BCD add/subtract extent/carry story is not complete.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/extent/extend/ ?

File: src/vec_bcd_ppc.h [hunk 2]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * There are still cases where the operation will generate a borrow
+ * and invert (10s complement) incorrectly.
+ * The net seems to be that for BCD multiple precision difference to
+ * work correctly, the larger magnitude must be the first
+ * operand.
+ *
+ * \subsubsection bcd128_extended_0_2_1 Vector BCD Multiply Quadword example
+ *
+ * TBD
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: is this "TBD" a keeper?

File: src/vec_bcd_ppc.h [hunk 3]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * in the current PowerISA. Decimal multiply and divide are
+ * supported in the Decimal Floating-point (DFP) Unit (DFU), as well
+ * as conversion to and from signed (unsigned) BCD.
+ *
+ * So BCD multiply and divide operations can be routed through the
+ * DFU with a few caveats.
+ * - DFP Extended format supports up to 34 digits precision
+ * - DFP significand represent digits to the <I>left</I> of the
+ * implied decimal point.
+ * - DFP finite number are not normalize.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/normalize/normalized/

File: src/vec_bcd_ppc.h [hunk 4]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ *
+ * The multiply case is bit more complicated as we need to produce up
+ * to 62 digit results without losing precision and DFP only supports
+ * 34 digits. This requires splitting the input operands into groups
+ * of digits where partial products of any combination of these groups
+ * is guaranteed not exceed 34 digits.
+ *
+ * One way to do this is split each 31-digit operand into two 16-digit
+ * chunks (actually 15 and 16-digits). These chunks are converted to
+ * DFP extended format and multiplied to product four 32-digit partial
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/product/produce/

File: src/vec_bcd_ppc.h [hunk 5]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * products. These partial products can be aligned and summed to
+ * produce the high  and low 31-digits of the full 62-digit product.
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: nit: extra space after "high"

File: src/vec_bcd_ppc.h [hunk 6]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * This is the basis for vec_bcd_mul(), vec_bcdmulh(), and
+ * vec_cbcdmul().
+ *
+ * A simple vec_and() can be used to isolate the low order 16 BCD
+ * digits. It is simple at this point to detect if both operands are
+ * 16-digits or less by comparing the original operand to isolate
+ * value. In this case the product can not exceed 32 digits and we
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/to isolate value/to the isolated value/ ?

File: src/vec_bcd_ppc.h [hunk 7]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+#ifdef _ARCH_PWR9
+  t = vec_bcdadd (t, _BCD_CONST_ZERO);
+#else
+  if (vec_all_eq((vui32_t) t, mz))
+    t = _BCD_CONST_ZERO;
+#endif
+  return t;
+ * \endcode
+ * From here the code diverges for multiply low and multiply high
+ * (and full combined multiple). Multiply low only needs the 3 lower
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/multiple/multiply/

File: src/vec_bcd_ppc.h [hunk 8]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * For this case use
+ * <B>DFP Shift Significand Left Immediate</B> and
+ * <B>DFP Shift Significand Right Immediate</B>. All the data is in
+ * the DFU and the high cost of the DFU shift is offset by avoiding
+ * extra format conversions. We use shift left 17 followed by shift
+ * right 1 to clear the highest order DFP digit and avoid any overflow.
+ * A final DFP add produces the low order 32 digits of the product
+ * which will be truncated to 31-digits in the conversion to BCD.
+ *
+ * How we can look at the BCD multiply high (generate the full 62-digit
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/How/Now/ ?

File: src/vec_bcd_ppc.h [hunk 9]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * product returning the high 31 digits) and point out the differences.
+ * Multiply high also start by isolating the low order 16 BCD
+ * digits, performing he low order multiply (low_a * low_b),
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/he/the/

File: src/vec_bcd_ppc.h [hunk 10]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+  else
+    {
+    ...
+    }
+  t = vec_DFP2BCD (d_t);
+ * \endcode
+ * So the short circuit code shifts the low partial product right 31
+ * digits and returns that value.
+ *
+ * If we can not short circuit Multiply high requires  the generation
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: need some separation after "short circuit", if I understand the sentence correctly.  Maybe simply a comma after "short circuit".

File: src/vec_bcd_ppc.h [hunk 11]
```diff
@@ -461,6 +556,244 @@ vec_cbcdaddcsq (vBCD_t *cout, vBCD_t a, vBCD_t b)
...
+ * \endcode
+ * Again we know that there are higher order digits in one or both
+ * operands and use vec_bcdsrqi() to isolate the high 15-digits of
+ * operands a and b. Then convert to DFP and multiply (high_a * low_b
+ * and high_b * low_a) for the two middle order partial products
+ * (d_hl and d_lh).
+ *
+ * The low order partial product (d_p) was generated above but we need
+ * only the high order 15 digits for summation. Shift the low partial
+ * produce right 16 digits then sum (d_hl + d_lh + d_ll) the low and
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: s/produce/product/

File: src/vec_bcd_ppc.h [hunk 12]
```diff
@@ -654,39 +995,36 @@ vec_bcdaddcsq (vBCD_t a, vBCD_t b)
...
+  const vui32_t mz = CONST_VINT128_W(0, 0, 0, 0x0000000d);
   d_a = vec_BCD2DFP (a);
   d_b = vec_BCD2DFP (b);
   d_s = d_a + d_b;
   // Shift right 31 digits, leaving the carry.
   d_t = __builtin_dscriq (d_s, 31);
   t = vec_DFP2BCD (d_t);
   // fix up spurious negative zeros
-  if (vec_all_eq((vui32_t) t, mz))
+  if (vec_all_eq((vui32_t ) t, mz))
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: extra space snuck in here.

File: src/vec_bcd_ppc.h [hunk 13]
```diff
@@ -1423,13 +2481,24 @@ vec_bcdsubcsq (vBCD_t a, vBCD_t b)
...
+#endif
+  a_b = vec_bcdsub (a, b);
   if (__builtin_expect (__builtin_bcdsub_ov ((vi128_t) a, (vi128_t) b, 0), 0))
     {
+#ifdef _ARCH_PWR9
+      t = vec_bcdcpsgn (_BCD_CONST_PLUS_ONE, a_b);
+#else
       if (__builtin_bcdsub_gt ((vi128_t) a, (vi128_t) b, 0))
-	t = _BCD_CONST_PLUS_ONE;
+      t = _BCD_CONST_PLUS_ONE;
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: not sure the indentation is right here

File: src/vec_bcd_ppc.h [hunk 14]
```diff
@@ -1423,13 +2481,24 @@ vec_bcdsubcsq (vBCD_t a, vBCD_t b)
...
       else
-	t = _BCD_CONST_MINUS_ONE;
+      t = _BCD_CONST_MINUS_ONE;
```
Comment (ThinkOpenly) [inline src/vec_bcd_ppc.h]: not sure the indentation is right here

File: src/vec_bcd_ppc.h [hunk 15]
```diff
@@ -30,8 +30,8 @@
 /*!
  * \file  vec_bcd_ppc.h
  * \brief Header package containing a collection of Binary Coded
- * Decimal (<B>BCD</B>) and Zoned Character computation operations
- * on vector registers.
+ * Decimal (<B>BCD</B>) computation  and Zoned Character conversion
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: ok

File: src/vec_bcd_ppc.h [hunk 16]
```diff
@@ -73,28 +53,21 @@
  * Densely packed decimal</A> (<I>DPD</I> and a more extensive set of
  * operations then BCD or Zoned. So DFP and the comprehensive C
  * language and runtime library support makes it a better target for
- * new business oriented applications.
+ * new business oriented applications. As the DFU supports conversions
+ * between DPD and BCD, existing DFP operations can be used emulate
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 17]
```diff
@@ -73,28 +53,21 @@
...
- * And conventions between _Float128 and _Decimal128 types is even
- * more challenging. Basically both POSIX and IEEE 754-2008 require
- * that it possible to convert floating-point values to an external
- * character decimal representation, with the specified rounding, and
- * back recovering the original value.
- * This always requires more precision for the conversion then is
- * available in the given format and size.
  *
- * So what are the operations we need, what does the PowerISA provide,
+ * So what operations are need, what does the PowerISA provide,
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 18]
```diff
@@ -168,7 +141,19 @@
  * POWER9 also added quadword binary multiply 10 with carry extend
  * forms than can also help with decimal to binary conversion.
  *
- * \note The compiler disables associated <altivec.h> built-ins if the
+ * The
+ * <A HREF="https://openpowerfoundation.org/?resource_lib=64-bit-elf-v2-abi-specification-power-architecture">
+ * OpenPOWER ABI</A> does have an
+ * <I>Appendix B. Binary-Coded Decimal Built-In Functions</I>
+ * and proposes that compiles provide a <B>bcd.h</B> header file.
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h [hunk 19]
```diff
@@ -216,11 +201,121 @@
...
+ * new business oriented applications.
+ * As DFP is supported directly in the hardware and has extensive
+ * language and runtime support, there is little that PVECLIB can
+ * contribute to general decimal radix computation.
+ *
+ * \note BCD and DFP support requires at least PowerISA 2.05 (POWER6)
+ * or later server level processor support.
+ *
+ * However the vector unit and recent BCD and Zoned extensions can
+ * still be useful in areas include large order multiple precision
```
[the remark this answers is not in the document]
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 1 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 2 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: Nop, removed

File: src/vec_bcd_ppc.h (hunk 3 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 4 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 5 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 6 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 7 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 8 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 9 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 10 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 11 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 12 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: OK

File: src/vec_bcd_ppc.h (hunk 13 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: Yea, This seems to be another case where the SDK GNU format tool not formating correctly

File: src/vec_bcd_ppc.h (hunk 14 above)
Comment (munroesj52) [reply src/vec_bcd_ppc.h]: Another one, Fixed

**Comment by munroesj52** [comment]
Thanks for the quick review. I suspect you know my pattern of misspelling and poor syntax by heart.

**Pull request merged**